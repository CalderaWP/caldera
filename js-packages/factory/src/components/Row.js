import React, {Fragment} from 'react';
import {Fields} from './Fields';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Flex, Box} from '@rebass/grid'


/**
 * Render a row of fields with a shared onChange/onBlur event for all values
 * @param columns
 * @param onChange
 * @param onBlur
 * @param className
 * @param rowId
 * @return {*}
 * @constructor
 */
export const Row = ({
						columns, onChange, onBlur, className, rowId
					}) => (
	<Flex
		className={classNames('caldera-row', className)}
		id={rowId}
	>
		{columns.map(column => {
			let {
				width,
				padding,
				columnId,
				fields,
			} = column;

			padding = padding ? padding : 8;

			return (
				<Box
					key={columnId}
					id={columnId}
					width={width}
					px={padding}
					py={padding}
				>
					<Fields
						fields={fields}
						onChange={onChange}
						onBlur={onBlur}
					/>
				</Box>

			)
		})}
	</Flex>

);

/**
 * Row prop types
 *
 * @type {{columns: shim, rowId: *, onChange: *, onBlur: shim, className: shim}}
 */
export const rowPropTypes = {
	columns: PropTypes.arrayOf(PropTypes.shape({
		fields: PropTypes.array.isRequired,
		width: PropTypes.string.isRequired,
		columnId: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number,
		])
		//input propTypes?
	})),
	rowId: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]),
	onChange: PropTypes.func.isRequired,
	onBlur: PropTypes.func,
	className: PropTypes.string
};
Row.propTypes = rowPropTypes;
