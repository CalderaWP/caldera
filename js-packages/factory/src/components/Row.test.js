import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField
} from '../fields.fixtures';

import {Row} from './Row';

describe('Row component', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn()
	});

	const columnWithEmailField = {
		fields:[
			emailField,
		],
		width: '1/2',
		columnId: 22,

	};

	const columnWithEmailFieldNoValue = {
		fields:[
			{...emailField,value: ''},
		],
		width: '1/2',
		columnId: 22,

	};

	const columnWithSelectField = {
		fields:[
			{...selectField, value: 2},
		],
		width: '1/2',
		columnId: 222,

	};

	const columns = [
		{
			fields:[
				textField,
				checkboxField,
			],
			width: '1/2',
			columnId: 1,
		},
		columnWithEmailField
	];


	it('Creates a 2 column  row', () => {

		const component = renderer.create(
			<Row
				rowId={'twoColumns'}
				columns={columns} onChange={onChange} onBlur={onBlur}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Calls the change handler correctly', () => {


		const component = mount(
			<Row
				rowId={'twoColumns'}
				columns={[columnWithEmailFieldNoValue]} onChange={onChange} onBlur={onBlur}
			/>
		);
		//emailField
		const expectedValue = 'roy@hiroy.club';
		const event = {target: {value: expectedValue}};
		component.find( 'input').simulate( 'change',event);
		expect(onChange.mock.calls[0][0]).toEqual({[emailField.fieldId]:expectedValue});



	});


});
