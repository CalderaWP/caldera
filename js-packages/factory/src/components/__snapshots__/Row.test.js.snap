// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row component Creates a 2 column  row 1`] = `
<div
  className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
  id="twoColumns"
>
  <div
    className="sc-bdVaJa dihzyz"
    id={1}
    width="1/2"
  >
    <div
      className="caldera-fields-wrapper"
    >
      <div
        className="components-base-control caldera-field caldera-field-text"
      >
        <div
          className="components-base-control__field"
        >
          <label
            className="components-base-control__label"
            htmlFor="inspector-text-control-0"
          >
            First Name
          </label>
          <input
            aria-describedby="inspector-text-control-0__help"
            className="components-text-control__input"
            id="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value="roy"
          />
        </div>
        <p
          className="components-base-control__help"
          id="inspector-text-control-0__help"
        >
          your first name
        </p>
      </div>
      <div
        className="components-base-control"
      >
        <div
          className="components-base-control__field"
        >
          <input
            aria-describedby="inspector-checkbox-control-0__help"
            checked={true}
            className="components-checkbox-control__input"
            id="checkboxFieldLabel"
            onChange={[Function]}
            type="checkbox"
            value="1"
          />
          <label
            className="components-checkbox-control__label"
            htmlFor="inspector-checkbox-control-0"
          >
            Checkbox Labe
          </label>
        </div>
        <p
          className="components-base-control__help"
          id="inspector-checkbox-control-0__help"
        >
          Checkbox field description
        </p>
      </div>
    </div>
  </div>
  <div
    className="sc-bdVaJa dihzyz"
    id={22}
    width="1/2"
  >
    <div
      className="caldera-fields-wrapper"
    >
      <div
        className="components-base-control caldera-field caldera-field-email"
      >
        <div
          className="components-base-control__field"
        >
          <label
            className="components-base-control__label"
            htmlFor="inspector-text-control-1"
          >
            Email Field Label
          </label>
          <input
            aria-describedby="inspector-text-control-1__help"
            className="components-text-control__input"
            id="emailFieldId"
            multiple={true}
            onBlur={[Function]}
            onChange={[Function]}
            placeholder={undefined}
            type="email"
            value="emai@email.com"
          />
        </div>
        <p
          className="components-base-control__help"
          id="inspector-text-control-1__help"
        >
          Email field description
        </p>
      </div>
    </div>
  </div>
</div>
`;
